native/vendor/google/billing/GoogleBilling.cpp

```diff
@@ -23,94 +23,282 @@
+    listeners->clear();
 }
 
-Purchase::~Purchase() {
-    GoogleBillingHelper::removePurchase(_id);
+BillingClient::~BillingClient() {
+    CC_ASSERT(_tag >= 0);
+    GoogleBillingHelper::removeGoogleBilling(_tag);
+    GoogleBillingManager::getInstance()->removeGoogleBilling(_tag);
+
+    DEL_JS_OBJ_REF(_purchasesUpdatedListener);
+    DEL_JS_OBJ_REF(_userChoiceBillingListener);
+    RemoveJsObject(&_billingClientStateListeners);
+    RemoveJsObject(&_productDetailsResponseListeners);
+    RemoveJsObject(&_consumeResponseListeners);
+    RemoveJsObject(&_acknowledgePurchaseResponseListeners);
+    RemoveJsObject(&_queryPurchasesResponseListeners);
+    RemoveJsObject(&_alternativeBillingOnlyReportingDetailsListeners);
+    RemoveJsObject(&_alternativeBillingOnlyAvailabilityListeners);
+    RemoveJsObject(&_externalOfferReportingDetailsListeners);
+    RemoveJsObject(&_externalOfferAvailabilityListeners);
+    RemoveJsObject(&_alternativeBillingOnlyInformationDialogListeners);
+    RemoveJsObject(&_externalOfferInformationDialogListeners);
+    RemoveJsObject(&_inappListeners);
+    RemoveJsObject(&_billingConfigListeners);
 }
 
-void GoogleBilling::startConnection() {
-    GoogleBillingHelper::startConnection();
+void BillingClient::startConnection(se::Object* listener) {
+    auto it = std::find(_billingClientStateListeners.begin(), _billingClientStateListeners.end(), listener);
+    if (it == _billingClientStateListeners.end()) {
+        listener->root();
+        listener->incRef();
+        _billingClientStateListeners.push_back(listener);
+        GoogleBillingHelper::startConnection(_tag, _billingClientStateListeners.size() - 1);
+    } else {
+        //
+    }
 }
 
-void GoogleBilling::endConnection() {
-    GoogleBillingHelper::endConnection();
+void BillingClient::endConnection() {
+    GoogleBillingHelper::endConnection(_tag);
 }
 
-int GoogleBilling::getConnectionState() const {
-    return GoogleBillingHelper::getConnectionState();
+int BillingClient::getConnectionState() const {
+    return GoogleBillingHelper::getConnectionState(_tag);
 }
 
-bool GoogleBilling::isReady() const {
-    return GoogleBillingHelper::isReady();
+bool BillingClient::isReady() const {
+    return GoogleBillingHelper::isReady(_tag);
 }
 
-void GoogleBilling::queryProductDetailsParams(const std::vector<std::string>& productIds, const std::string& type) {
-    GoogleBillingHelper::queryProductDetailsParams(productIds, type);
+void BillingClient::queryProductDetailsAsync(QueryProductDetailsParams* params, se::Object* listener) {
+    auto it = std::find(_productDetailsResponseListeners.begin(), _productDetailsResponseListeners.end(), listener);
+    if (it == _productDetailsResponseListeners.end()) {
+        listener->root();
+        listener->incRef();
+        _productDetailsResponseListeners.push_back(listener);
+        std::vector<std::string> productIDs;
+        std::vector<std::string> productTypes;

```

Comment:
Could reserve the capacity for these two `std::vector` to avoid memory re-allocation inside `std::vector`.
```suggestion
        std::vector<std::string> productTypes;
        productTypes.reserve(params->_productList.size());
```

Reply:
fixed.